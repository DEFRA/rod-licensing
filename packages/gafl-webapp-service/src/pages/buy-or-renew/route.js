import { BUY_OR_RENEW } from '../../uri.js'
import Joi from 'joi'
import pageRoute from '../../routes/page-route.js'
import { nextPage } from '../../routes/next-page.js'
export const buyOrRenew = {
  buy: 'buy',
  renew: 'renew'
}

export const validator = Joi.object({
  'buy-or-renew': Joi.string()
    .valid('buy-licence', 'renew-licence')
    .required()
}).options({ abortEarly: false, allowUnknown: true })

const route = pageRoute(BUY_OR_RENEW.page, BUY_OR_RENEW.uri, validator, nextPage)

export default route
