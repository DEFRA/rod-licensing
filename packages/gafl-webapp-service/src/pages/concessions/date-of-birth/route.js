import { DATE_OF_BIRTH } from '../../../uri.js'
import Joi from 'joi'
import pageRoute from '../../../routes/page-route.js'
import { validation } from '@defra-fish/business-rules-lib'
import { nextPage } from '../../../routes/next-page.js'

export const validator = payload => {
  const dateOfBirth = `${payload['date-of-birth-year']}-${payload['date-of-birth-month']}-${payload['date-of-birth-day']}`
  Joi.assert(
    { 'date-of-birth': dateOfBirth },
    Joi.object({
      'date-of-birth': validation.contact.createBirthDateValidator(Joi)
    })
  )
}

export const getData = async request => {
  const { isLicenceForYou } = await request.cache().helpers.status.getCurrentPermission()

  return { isLicenceForYou }
}

export default pageRoute(DATE_OF_BIRTH.page, DATE_OF_BIRTH.uri, validator, nextPage, getData)
