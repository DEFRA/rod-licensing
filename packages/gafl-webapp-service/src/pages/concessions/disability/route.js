import { DISABILITY_CONCESSION } from '../../../uri.js'
import pageRoute from '../../../routes/page-route.js'
import Joi from 'joi'
import { validation } from '@defra-fish/business-rules-lib'
import * as concessionHelper from '../../../processors/concession-helper.js'
import { nextPage } from '../../../routes/next-page.js'
import { disabilityConcessionTypes } from './update-transaction.js'

const validator = Joi.object({
  'disability-concession': Joi.string()
    .valid(...Object.values(disabilityConcessionTypes))
    .required(),
  'ni-number': Joi.alternatives().conditional('disability-concession', {
    is: disabilityConcessionTypes.pipDla,
    then: validation.contact.createNationalInsuranceNumberValidator(Joi).required(),
    otherwise: Joi.string().empty('')
  }),
  'blue-badge-number': Joi.alternatives().conditional('disability-concession', {
    is: disabilityConcessionTypes.blueBadge,
    then: Joi.string()
      .max(25)
      .required(),
    otherwise: Joi.string().empty('')
  })
}).options({ abortEarly: false, allowUnknown: true })

export const getData = async request => {
  const permission = await request.cache().helpers.transaction.getCurrentPermission()

  return {
    hasJunior: concessionHelper.hasJunior(permission),
    hasSenior: concessionHelper.hasSenior(permission),
    isLicenceForYou: permission.isLicenceForYou,
    ...disabilityConcessionTypes
  }
}

export default pageRoute(DISABILITY_CONCESSION.page, DISABILITY_CONCESSION.uri, validator, nextPage, getData)
