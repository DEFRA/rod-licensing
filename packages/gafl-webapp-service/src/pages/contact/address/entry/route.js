import { ADDRESS_ENTRY, ADDRESS_LOOKUP } from '../../../../uri.js'
import pageRoute from '../../../../routes/page-route.js'
import Joi from 'joi'
import { validation } from '@defra-fish/business-rules-lib'
import { countries } from '../../../../processors/refdata-helper.js'
import { nextPage } from '../../../../routes/next-page.js'
import { getPronoun } from '../../../../processors/licence-type-display.js'

export const validator = Joi.object({
  premises: validation.contact.createPremisesValidator(Joi),
  street: validation.contact.createStreetValidator(Joi),
  locality: validation.contact.createLocalityValidator(Joi),
  town: validation.contact.createTownValidator(Joi),
  postcode: Joi.alternatives().conditional('country-code', {
    is: Joi.string().pattern(/^GB/),
    then: validation.contact.createUKPostcodeValidator(Joi),
    otherwise: validation.contact.createOverseasPostcodeValidator(Joi)
  }),
  'country-code': Joi.string().required()
}).options({ abortEarly: false, allowUnknown: true })

export const getCountryDropDownOptions = async () => {
  const options = await countries.getAll()
  return options.filter(country => country.code !== 'GB')
}

export const getData = async request => {
  const { addresses, searchTerms } = await request.cache().helpers.addressLookup.getCurrentPermission()
  const { isLicenceForYou } = await request.cache().helpers.transaction.getCurrentPermission()

  const pronoun = getPronoun(isLicenceForYou).possessive

  return {
    pronoun,
    searchTerms: !addresses?.length && searchTerms ? searchTerms : null,
    countries: await getCountryDropDownOptions(),
    lookupPage: ADDRESS_LOOKUP.uri
  }
}

export default pageRoute(ADDRESS_ENTRY.page, ADDRESS_ENTRY.uri, validator, nextPage, getData)
