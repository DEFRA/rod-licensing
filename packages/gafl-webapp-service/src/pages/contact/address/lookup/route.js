import { ADDRESS_LOOKUP, ADDRESS_ENTRY, OS_TERMS } from '../../../../uri.js'
import pageRoute from '../../../../routes/page-route.js'
import Joi from 'joi'
import { hasJunior, validation } from '@defra-fish/business-rules-lib'
import { isPhysical } from '../../../../processors/licence-type-display.js'
import { nextPage } from '../../../../routes/next-page.js'
import { addLanguageCodeToUri } from '../../../../processors/uri-helper.js'

const validator = Joi.object({
  premises: validation.contact.createPremisesValidator(Joi),
  postcode: validation.contact.createUKPostcodeValidator(Joi)
}).options({ abortEarly: false, allowUnknown: true })

export const getData = async request => {
  const permission = await request.cache().helpers.transaction.getCurrentPermission()

  return {
    isLicenceForYou: permission.isLicenceForYou,
    licenceLength: permission.licenceLength,
    junior: hasJunior(permission),
    isPhysical: isPhysical(permission),
    uri: {
      entryPage: addLanguageCodeToUri(request, ADDRESS_ENTRY.uri),
      osTerms: addLanguageCodeToUri(request, OS_TERMS.uri)
    }
  }
}

export default pageRoute(ADDRESS_LOOKUP.page, ADDRESS_LOOKUP.uri, validator, nextPage, getData)
