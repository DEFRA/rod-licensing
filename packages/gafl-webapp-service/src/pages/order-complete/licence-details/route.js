import Boom from '@hapi/boom'

import pageRoute from '../../../routes/page-route.js'
import { LICENCE_DETAILS } from '../../../uri.js'
import { COMPLETION_STATUS } from '../../../constants.js'
import { nextPage } from '../../../routes/next-page.js'
import { licenceTypeDisplay } from '../../../processors/licence-type-display.js'
import { displayStartTime, displayEndTime } from '../../../processors/date-and-time-display.js'
import * as concessionHelper from '../../../processors/concession-helper.js'

const getAgeConcessionText = (permission, catalog) => {
  if (concessionHelper.hasSenior(permission)) {
    return catalog.age_senior_concession
  }
  if (concessionHelper.hasJunior(permission)) {
    return catalog.age_junior_concession
  }
}

export const getData = async request => {
  const status = await request.cache().helpers.status.get()

  if (!status[COMPLETION_STATUS.agreed]) {
    throw Boom.forbidden('Attempt to access the licence information handler with no agreed flag set')
  }

  if (!status[COMPLETION_STATUS.posted]) {
    throw Boom.forbidden('Attempt to access the licence information handler with no posted flag set')
  }

  if (!status[COMPLETION_STATUS.finalised]) {
    throw Boom.forbidden('Attempt to access the licence information handler with no finalised flag set')
  }

  const permission = await request.cache().helpers.transaction.getCurrentPermission()

  const startTimeString = displayStartTime(request, permission)
  const endTimeString = displayEndTime(request, permission)
  const catalog = request.i18n.getCatalog()

  return {
    permission,
    startTimeString,
    endTimeString,
    disabled: concessionHelper.hasDisabled(permission),
    ageConcession: concessionHelper.getAgeConcession(permission),
    ageConcessionText: getAgeConcessionText(permission, catalog),
    licenceTypeStr: licenceTypeDisplay(permission, catalog)
  }
}

export default pageRoute(LICENCE_DETAILS.page, LICENCE_DETAILS.uri, null, nextPage, getData)
