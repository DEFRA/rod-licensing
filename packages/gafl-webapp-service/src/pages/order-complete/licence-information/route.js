import Boom from '@hapi/boom'

import pageRoute from '../../../routes/page-route.js'
import { LICENCE_INFORMATION } from '../../../uri.js'
import { COMPLETION_STATUS } from '../../../constants.js'
import { nextPage } from '../../../routes/next-page.js'
import { licenceTypeDisplay } from '../../../processors/licence-type-display.js'
import { displayStartTime, displayEndTime } from '../../../processors/date-and-time-display.js'
import * as concessionHelper from '../../../processors/concession-helper.js'

const getData = async request => {
  const status = await request.cache().helpers.status.get()

  if (!status[COMPLETION_STATUS.agreed]) {
    throw Boom.forbidden('Attempt to access the licence information handler with no agreed flag set')
  }

  if (!status[COMPLETION_STATUS.posted]) {
    throw Boom.forbidden('Attempt to access the licence information handler with no posted flag set')
  }

  if (!status[COMPLETION_STATUS.finalised]) {
    throw Boom.forbidden('Attempt to access the licence information handler with no finalised flag set')
  }

  const permission = await request.cache().helpers.transaction.getCurrentPermission()

  const startTimeString = displayStartTime(permission)
  const endTimeString = displayEndTime(permission)

  return {
    permission,
    startTimeString,
    endTimeString,
    disabled: concessionHelper.hasDisabled(permission),
    licenceTypeStr: licenceTypeDisplay(permission)
  }
}

export default pageRoute(LICENCE_INFORMATION.page, LICENCE_INFORMATION.uri, null, nextPage, getData)
