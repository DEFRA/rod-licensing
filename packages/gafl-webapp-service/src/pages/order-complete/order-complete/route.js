import pageRoute from '../../../routes/page-route.js'

import Boom from '@hapi/boom'
import { COMPLETION_STATUS, FEEDBACK_URI_DEFAULT } from '../../../constants.js'
import { ORDER_COMPLETE, NEW_TRANSACTION, LICENCE_INFORMATION } from '../../../uri.js'
import { displayStartTime } from '../../../processors/date-and-time-display.js'
import * as mappings from '../../../processors/mapping-constants.js'
import { nextPage } from '../../../routes/next-page.js'

const getData = async request => {
  const status = await request.cache().helpers.status.get()
  const permission = await request.cache().helpers.transaction.getCurrentPermission()

  // If the agreed flag is not set to true then throw an exception
  if (!status[COMPLETION_STATUS.agreed]) {
    throw Boom.forbidden('Attempt to access the completion page handler with no agreed flag set')
  }

  // If the agreed flag is not set to true then throw an exception
  if (!status[COMPLETION_STATUS.posted]) {
    throw Boom.forbidden('Attempt to access the completion page handler with no posted flag set')
  }

  // If the finalised flag has not been set throw an exception
  if (!status[COMPLETION_STATUS.finalised]) {
    throw Boom.forbidden('Attempt to access the completion page handler with no finalised flag set')
  }

  await request.cache().helpers.status.set({ [COMPLETION_STATUS.completed]: true })
  await request.cache().helpers.status.setCurrentPermission({ currentPage: ORDER_COMPLETE.page })

  const startTimeStringTitle = displayStartTime(permission)
  return {
    permission,
    startTimeStringTitle,
    licenceTypes: mappings.LICENCE_TYPE,
    isPostalFulfilment: permission.licensee.postalFulfilment,
    contactMethod: permission.licensee.preferredMethodOfConfirmation,
    howContacted: mappings.HOW_CONTACTED,
    uri: {
      new: NEW_TRANSACTION.uri,
      feedback: process.env.FEEDBACK_URI || FEEDBACK_URI_DEFAULT,
      licenceInformation: LICENCE_INFORMATION.uri
    }
  }
}

export default pageRoute(ORDER_COMPLETE.page, ORDER_COMPLETE.uri, null, nextPage, getData)
