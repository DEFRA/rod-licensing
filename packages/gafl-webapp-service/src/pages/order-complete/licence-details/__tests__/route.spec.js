import { getData } from '../route'
import Boom from '@hapi/boom'
import { displayStartTime, displayEndTime } from '../../../../processors/date-and-time-display.js'
import { licenceTypeDisplay, licenceTypeAndLengthDisplay } from '../../../../processors/licence-type-display.js'
import { COMPLETION_STATUS, CommonResults, ShowDigitalLicencePages } from '../../../../constants.js'
import * as concessionHelper from '../../../../processors/concession-helper.js'

jest.mock('../../../../processors/concession-helper.js')
jest.mock('../../../../processors/licence-type-display.js')
jest.mock('../../../../processors/date-and-time-display.js')
jest.mock('../../../../constants', () => ({
  COMPLETION_STATUS: {
    agreed: 'agreed',
    posted: 'posted',
    finalised: 'finalised'
  },
  CommonResults: {
    ok: 'ok'
  },
  ShowDigitalLicencePages: {
    YES: 'yes'
  },
  MultibuyForYou: {
    YES: 'yes',
    NO: 'no'
  }
}))

const status = () => ({
  [COMPLETION_STATUS.agreed]: 'agreed',
  [COMPLETION_STATUS.posted]: 'posted',
  [COMPLETION_STATUS.finalised]: 'finalised',
  [CommonResults.OK]: 'ok',
  [ShowDigitalLicencePages.YES]: 'yes'
})

const transaction = () => ({
  permissions: [permission]
})

const permission = {
  licensee: {
    firstName: 'Turanga',
    lastName: 'Leela',
    obfuscatedDob: 'obfuscatedDob'
  },
  referenceNumber: '123456789',
  licenceType: 'trout-and-coarse',
  numberOfRods: '2',
  licenceLength: '8D',
  permit: { cost: 12 }
}

const catalog = Symbol('mock catalog')

describe('licence-length > route', () => {
  beforeEach(() => jest.clearAllMocks())

  describe('getData', () => {
    const getMockRequest = (statusGet = () => {}, transactionGet = () => {}) => ({
      cache: () => ({
        helpers: {
          transaction: {
            get: transactionGet
          },
          status: {
            get: statusGet
          }
        }
      }),
      i18n: {
        getCatalog: () => catalog
      }
    })

    it('licenceTypeDisplay is called with the expected arguments', async () => {
      await getData(getMockRequest(status, transaction))
      expect(licenceTypeDisplay).toHaveBeenCalledWith(permission, catalog)
    })

    it('licenceTypeAndLengthDisplay is called with the expected arguments', async () => {
      await getData(getMockRequest(status, transaction))
      expect(licenceTypeAndLengthDisplay).toHaveBeenCalledWith(permission, catalog)
    })

    it('displayStartTime is called with the expected arguments', async () => {
      const mockRequest = getMockRequest(status, transaction)
      await getData(mockRequest)
      expect(displayStartTime).toHaveBeenCalledWith(mockRequest, permission)
    })

    it('displayEndTime is called with the expected arguments', async () => {
      const mockRequest = getMockRequest(status, transaction)
      await getData(mockRequest)
      expect(displayEndTime).toHaveBeenCalledWith(mockRequest, permission)
    })

    it('hasDisabled is called with the expected arguments', async () => {
      await getData(getMockRequest(status, transaction))
      expect(concessionHelper.hasDisabled).toHaveBeenCalledWith(permission)
    })

    it('getAgeConcession is called with the expected arguments', async () => {
      await getData(getMockRequest(status, transaction))
      expect(concessionHelper.getAgeConcession).toHaveBeenCalledWith(permission)
    })

    describe('returns expected licence data for the given permission:', () => {
      const licenceType = Symbol('Trout and coarse, up to 2 rods')
      licenceTypeDisplay.mockReturnValue(licenceType)

      const licenceLength = Symbol('8 days')
      licenceTypeAndLengthDisplay.mockReturnValue(licenceLength)

      const startTime = Symbol('9:32am on 23 June 2021')
      displayStartTime.mockReturnValue(startTime)

      const endTime = Symbol('9:32am on 23 June 2022')
      displayEndTime.mockReturnValue(endTime)

      const disabled = Symbol('disabled')
      concessionHelper.hasDisabled.mockReturnValue(disabled)

      const concession = Symbol('concession')
      concessionHelper.getAgeConcession.mockReturnValue(concession)

      it('referenceNumber', async () => {
        const data = await getData(getMockRequest(status, transaction))
        expect(data.licences[0].referenceNumber).toBe('123456789')
      })

      it('licenceHolder', async () => {
        const data = await getData(getMockRequest(status, transaction))
        expect(data.licences[0].licenceHolder).toBe('Turanga Leela')
      })

      it('obfuscatedDob', async () => {
        const data = await getData(getMockRequest(status, transaction))
        expect(data.licences[0].obfuscatedDob).toBe('obfuscatedDob')
      })

      it('type', async () => {
        const data = await getData(getMockRequest(status, transaction))
        expect(data.licences[0].type).toBe(licenceType)
      })

      it('length', async () => {
        const data = await getData(getMockRequest(status, transaction))
        expect(data.licences[0].length).toBe(licenceLength)
      })

      it('start', async () => {
        const data = await getData(getMockRequest(status, transaction))
        expect(data.licences[0].start).toBe(startTime)
      })

      it('end', async () => {
        const data = await getData(getMockRequest(status, transaction))
        expect(data.licences[0].end).toBe(endTime)
      })

      it('price', async () => {
        const data = await getData(getMockRequest(status, transaction))
        expect(data.licences[0].price).toBe(permission.permit.cost)
      })

      it('disabled', async () => {
        const data = await getData(getMockRequest(status, transaction))
        expect(data.licences[0].disabled).toBe(disabled)
      })

      it('ageConcession', async () => {
        const data = await getData(getMockRequest(status, transaction))
        expect(data.licences[0].ageConcession).toBe(concession)
      })

      it('index', async () => {
        const data = await getData(getMockRequest(status, transaction))
        expect(data.licences[0].index).toBe(0)
      })

      it('returns expected data', async () => {
        const data = await getData(getMockRequest(status, transaction))
        expect(data).toMatchSnapshot()
      })
    })

    describe('throws a Boom forbidden error', () => {
      it('if status agreed flag is not set', async () => {
        const status = () => ({})
        const boomError = Boom.forbidden('Attempt to access the licence information handler with no agreed flag set')
        await expect(getData(getMockRequest(status, transaction))).rejects.toThrow(boomError)
      })

      it('if status posted flag is not set', async () => {
        const status = () => ({
          [COMPLETION_STATUS.agreed]: 'agreed'
        })
        const boomError = Boom.forbidden('Attempt to access the licence information handler with no posted flag set')
        await expect(getData(getMockRequest(status, transaction))).rejects.toThrow(boomError)
      })

      it('if status finalised flag is not set', async () => {
        const status = () => ({
          [COMPLETION_STATUS.agreed]: 'agreed',
          [COMPLETION_STATUS.posted]: 'posted'
        })
        const boomError = Boom.forbidden('Attempt to access the licence information handler with no finalised flag set')
        await expect(getData(getMockRequest(status, transaction))).rejects.toThrow(boomError)
      })
    })
  })
})
