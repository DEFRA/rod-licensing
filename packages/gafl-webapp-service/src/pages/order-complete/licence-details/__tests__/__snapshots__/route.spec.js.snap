// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`licence-length > route getData returns expected licence data for the given permission: returns expected data 1`] = `
Object {
  "licences": Array [
    Object {
      "ageConcession": Symbol(concession),
      "disabled": Symbol(disabled),
      "end": Symbol(9:32am on 23 June 2022),
      "index": 0,
      "length": Symbol(8 days),
      "licenceHolder": "Turanga Leela",
      "obfuscatedDob": "obfuscatedDob",
      "price": 12,
      "referenceNumber": "123456789",
      "start": Symbol(9:32am on 23 June 2021),
      "type": Symbol(Trout and coarse, up to 2 rods),
    },
  ],
}
`;
